File: src/niepce/ui/gridviewmodule.cpp
/*
 * niepce - ui/gridviewmodule.cpp
 *
 * Copyright (C) 2009-2022 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <gtkmm/popovermenu.h>

#include <exempi/xmpconsts.h>

#include "rust_bindings.hpp"

#include "fwk/base/debug.hpp"
#include "fwk/toolkit/application.hpp"
#include "fwk/toolkit/configdatabinder.hpp"
#include "fwk/toolkit/widgets/dock.hpp"
#include "gridviewmodule.hpp"

#include "rust_bindings.hpp"

namespace ui {

std::shared_ptr<GridViewModule> grid_view_module_new(const ui::SelectionController& selection_controller,
                                                     const GMenu* menu_, const eng::LibraryClientHost& client_host)
{
    Glib::RefPtr<Gio::Menu> menu;
    if (menu_) {
        menu = Glib::wrap(const_cast<GMenu*>(menu_));
    }
    return std::make_shared<GridViewModule>(selection_controller, menu, client_host);
}

GridViewModule::GridViewModule(const ui::SelectionController& selection_controller,
                               Glib::RefPtr<Gio::Menu> menu, const eng::LibraryClientHost& client_host)
  : m_selection_controller(selection_controller)
  , m_menu(menu)
  , m_libclient_host(client_host)
  , m_librarylistview(nullptr)
  , m_lib_splitview(Gtk::Orientation::HORIZONTAL)
  , m_dock(nullptr)
  , m_context_menu(nullptr)
{
}

GridViewModule::~GridViewModule()
{
    m_widget = nullptr;
}

void
GridViewModule::on_lib_notification(const eng::LibNotification &ln, const eng::LibraryClientWrapper& client) const
{
    switch (ln.type_()) {
    case eng::NotificationType::METADATA_QUERIED:
    {
        auto& lm = ln.get_libmetadata();
        DBG_OUT("received metadata");
        m_metapanecontroller->display(lm.id(), &lm);
        break;
    }
    case eng::NotificationType::METADATA_CHANGED:
    {
        DBG_OUT("metadata changed");
        auto id = ln.id();
        if(id && id == m_metapanecontroller->displayed_file()) {
            // FIXME: actually just update the metadata
            client.request_metadata(id);
        }
        break;
    }
    default:
        break;
    }
}

void GridViewModule::display_none() const
{
    m_metapanecontroller->display(0, nullptr);
}

Gtk::Widget * GridViewModule::buildWidget()
{
  if(m_widget) {
    return m_widget;
  }
  m_widget = &m_lib_splitview;
  m_context_menu = Gtk::manage(new Gtk::PopoverMenu(m_menu));
  auto& model = m_selection_controller.get_list_store();

  auto image_grid_view = npc::npc_image_grid_view_new(
      GTK_SINGLE_SELECTION(model.unwrap_ref().gobj()),
      GTK_POPOVER_MENU(m_context_menu->gobj()),
      m_libclient_host
  );
  m_librarylistview = Gtk::manage(Glib::wrap(image_grid_view->get_grid_view()));
  m_image_grid_view = std::move(image_grid_view);
  m_librarylistview->set_vexpand(true);

  m_context_menu->set_parent(*m_librarylistview);
  m_context_menu->set_has_arrow(false);
  m_librarylistview->signal_unrealize().connect(
      sigc::mem_fun(*m_context_menu, &Gtk::PopoverMenu::unparent));

  auto gesture = Gtk::GestureClick::create();
  m_librarylistview->add_controller(gesture);
  gesture->signal_pressed().connect([this, gesture] (int, double x, double y) {
      this->on_librarylistview_click(gesture, x, y);
  });

  m_image_grid_view.value()->add_rating_listener(
      std::make_unique<npc::RatingClickListener>([this] (int64_t id, int32_t rating) {
          DBG_OUT("rating changed %ld %d", id, rating);
          this->on_rating_changed(id, rating);
      }));

  m_scrollview.set_child(*m_librarylistview);
  m_scrollview.set_policy(Gtk::PolicyType::AUTOMATIC, Gtk::PolicyType::AUTOMATIC);
  m_lib_splitview.set_wide_handle(true);

  // build the toolbar
  auto box = Gtk::manage(new Gtk::Box(Gtk::Orientation::VERTICAL));
  box->append(m_scrollview);
  auto toolbar = ui::image_toolbar_new();
  gtk_box_append(box->gobj(), GTK_WIDGET(toolbar));
  m_lib_splitview.set_start_child(*box);

  m_dock = new fwk::Dock();
  m_metapanecontroller = MetaDataPaneController::Ptr(new MetaDataPaneController);
  m_metapanecontroller->signal_metadata_changed.connect(
      sigc::mem_fun(*this, &GridViewModule::on_metadata_changed));
  add(m_metapanecontroller);
  m_lib_splitview.set_end_child(*m_dock);
  m_dock->vbox().append(*m_metapanecontroller->buildWidget());

  m_databinders.add_binder(new fwk::ConfigDataBinder<int>(
                             m_lib_splitview.property_position(),
                             fwk::Application::app()->config(),
                             "meta_pane_splitter"));
  return m_widget;
}

void GridViewModule::dispatch_action(const std::string & /*action_name*/)
{
}

Gtk::GridView * GridViewModule::image_list() const
{
    return m_librarylistview;
}

void GridViewModule::on_metadata_changed(const fwk::WrappedPropertyBagPtr& props,
                                         const fwk::WrappedPropertyBagPtr& old)
{
    // TODO this MUST be more generic
    DBG_OUT("on_metadata_changed()");
    m_selection_controller.set_properties(*props, *old);
}

void GridViewModule::on_rating_changed(eng::library_id_t id, int32_t rating) const
{
    m_selection_controller.set_rating_of(id, rating);
}

void GridViewModule::on_librarylistview_click(const Glib::RefPtr<Gtk::GestureClick>& gesture, double x, double y)
{
    auto button = gesture->get_current_button();
    DBG_OUT("GridView click handler, button: %u", button);
    if (button == 3 && !m_librarylistview->get_model()->get_selection()->is_empty()) {
        m_context_menu->set_pointing_to(Gdk::Rectangle(x, y, 1, 1));
        m_context_menu->popup();

        return;
    }
}

}
/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  c-basic-offset:4
  tab-width:4
  indent-tabs-mode:nil
  fill-column:80
  End:
*/

File: niepce-main/src/niepce/ui/grid_view_module.rs
/*
 * niepce - niepce/ui/grid_view_module.rs
 *
 * Copyright (C) 2022-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::rc::Rc;
use std::sync::Arc;

use gtk4::prelude::*;
use npc_fwk::{gdk4, gio, glib, gtk4};

use npc_engine::catalog;
use npc_engine::library::notification::LibNotification;
use npc_engine::libraryclient::{ClientInterface, LibraryClient, LibraryClientHost};
use npc_fwk::toolkit::widgets::Dock;
use npc_fwk::toolkit::widgets::MetadataPropertyBag;
use npc_fwk::toolkit::{Controller, ControllerImplCell, UiController};
use npc_fwk::{dbg_out, send_async_local};

use crate::niepce::ui::metadata_pane_controller::MetadataOutputMsg;
use crate::niepce::ui::{
    ImageGridView, LibraryModule, MetadataPaneController, SelectionController,
};

pub enum GridMsg {
    Click(gtk4::GestureClick, f64, f64),
    ChangeRating(catalog::LibraryId, i32),
    MetadataChanged(MetadataPropertyBag, MetadataPropertyBag),
}

pub struct GridViewModule {
    imp_: ControllerImplCell<GridMsg, ()>,
    selection_controller: Rc<SelectionController>,
    pub image_grid_view: ImageGridView,
    metadatapanecontroller: Rc<MetadataPaneController>,
    context_menu: gtk4::PopoverMenu,
    widget: gtk4::Paned,
}

impl Controller for GridViewModule {
    type InMsg = GridMsg;
    type OutMsg = ();

    npc_fwk::controller_imp_imp!(imp_);

    fn dispatch(&self, msg: GridMsg) {
        match msg {
            GridMsg::Click(gesture, x, y) => self.on_librarylistview_click(&gesture, x, y),
            GridMsg::ChangeRating(id, rating) => {
                self.selection_controller.set_rating_of(id, rating)
            }
            GridMsg::MetadataChanged(new, old) => {
                self.selection_controller.set_properties(&new, &old)
            }
        }
    }
}

impl UiController for GridViewModule {
    fn widget(&self) -> &gtk4::Widget {
        // In this the assumption is that widget has been set at
        // construction time from the C++ impl and since the only way
        // to do so is by calling the new associated function, it
        // should be ok.
        self.widget.upcast_ref()
    }

    fn actions(&self) -> Option<(&str, &gio::ActionGroup)> {
        None
    }
}

impl LibraryModule for GridViewModule {
    fn widget(&self) -> &gtk4::Widget {
        UiController::widget(self)
    }
}

impl GridViewModule {
    pub fn new(
        selection_controller: &Rc<SelectionController>,
        menu: &gio::Menu,
        libclient_host: &LibraryClientHost,
    ) -> Rc<Self> {
        let widget = gtk4::Paned::new(gtk4::Orientation::Horizontal);
        let context_menu = gtk4::PopoverMenu::from_model(Some(menu));
        let model = selection_controller.list_store().selection_model();
        let image_grid_view = ImageGridView::new(
            model.clone(),
            Some(context_menu.clone()),
            Some(libclient_host.shared_ui_provider()),
        );
        let metadatapanecontroller = MetadataPaneController::new();
        let mut module = GridViewModule {
            imp_: ControllerImplCell::default(),
            selection_controller: selection_controller.clone(),
            context_menu,
            image_grid_view,
            metadatapanecontroller,
            widget,
        };

        module.build_widget();

        let module = Rc::new(module);
        <Self as Controller>::start(&module);

        module
    }

    fn build_widget(&mut self) {
        self.image_grid_view.set_vexpand(true);
        self.context_menu.set_parent(&*self.image_grid_view);
        self.context_menu.set_has_arrow(false);
        self.image_grid_view.connect_unrealize(glib::clone!(
            #[strong(rename_to = menu)]
            self.context_menu,
            move |_| {
                menu.unparent();
            }
        ));

        let gesture = gtk4::GestureClick::new();
        self.image_grid_view.add_controller(gesture.clone());
        let sender = self.sender();
        gesture.connect_pressed(glib::clone!(
            #[strong]
            sender,
            move |gesture, _, x, y| {
                let gesture = gesture.clone();
                send_async_local!(GridMsg::Click(gesture, x, y), sender);
            }
        ));
        let sender = self.sender();
        self.image_grid_view
            .add_rating_listener(Box::new(move |(id, rating)| {
                send_async_local!(GridMsg::ChangeRating(id, rating), sender);
            }));

        let scrollview = gtk4::ScrolledWindow::new();
        scrollview.set_child(Some(&*self.image_grid_view));
        scrollview.set_policy(gtk4::PolicyType::Automatic, gtk4::PolicyType::Automatic);
        self.widget.set_wide_handle(true);
        let box_ = gtk4::Box::new(gtk4::Orientation::Vertical, 0);
        box_.append(&scrollview);
        let toolbar = crate::niepce::ui::imagetoolbar::image_toolbar_new();
        box_.append(&toolbar);
        self.widget.set_start_child(Some(&box_));

        let dock = Dock::new();
        let sender = self.sender();
        self.metadatapanecontroller
            .set_forwarder(Some(Box::new(move |msg| match msg {
                MetadataOutputMsg::MetadataChanged(new, old) => {
                    send_async_local!(GridMsg::MetadataChanged(new, old), sender)
                }
            })));

        self.widget.set_end_child(Some(&dock));
        self.widget.set_resize_end_child(false);
        dock.vbox().append(self.metadatapanecontroller.widget());
    }

    fn on_librarylistview_click(&self, gesture: &gtk4::GestureClick, x: f64, y: f64) {
        let button = gesture.current_button();
        dbg_out!("GridView click handler, button: {button}");
        if button == 3
            && !self
                .image_grid_view
                .model()
                .map(|model| model.selection().is_empty())
                .unwrap_or(true)
        {
            self.context_menu
                .set_pointing_to(Some(&gdk4::Rectangle::new(x as i32, y as i32, 1, 1)));
            self.context_menu.popup();
        }
    }

    pub fn on_lib_notification(&self, ln: &LibNotification, client: &Arc<LibraryClient>) {
        match ln {
            LibNotification::MetadataQueried(lm) => {
                self.metadatapanecontroller.display(lm.id(), Some(lm));
            }
            LibNotification::MetadataChanged(lm) => {
                if lm.id != 0 && self.metadatapanecontroller.displayed() == lm.id {
                    client.request_metadata(lm.id);
                }
            }
            _ => (),
        }
    }

    pub fn display_none(&self) {
        self.metadatapanecontroller.display(0, None);
    }
}
